spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveIcebergProcedures.scala

```diff
@@ -0,0 +1,132 @@
+      val values = buildProcedureValues(procedure, args)
+      procedure.createRunnableCommand(values)
+  }
+
+  private def buildProcedureValues(procedure: Procedure, args: Seq[CallArgument]): Array[Any] = {
+    // build a map of declared parameter names to their positions
+    val params = procedure.parameters
+    val nameToPositionMap = params.map(_.name).zipWithIndex.toMap
+
+    // verify named and positional args are not mixed
+    val containsNamedArg = args.exists(arg => arg.name.isDefined)
+    val containsPositionalArg = args.exists(arg => arg.name.isEmpty)
+    if (containsNamedArg && containsPositionalArg) {
+      throw new AnalysisException("Named and positional arguments cannot be mixed")
+    }
+
+    // build a map of parameter names to args
+    val nameToArgMap = if (containsNamedArg) {
+      buildNameToArgMapUsingNames(args, nameToPositionMap)
+    } else {
+      buildNameToArgMapUsingPositions(args, params)
+    }
+
+    // verify all required parameters are provided
+    params.filter(_.required)
+      .find(param => !nameToArgMap.contains(param.name))
+      .foreach { missingArg =>
+        throw new AnalysisException(s"Required procedure argument '${missingArg.name}' is missing")
+      }
+
+    val values = new Array[Any](params.size)
+
+    // convert provided args from internal Spark representation to Scala
+    nameToArgMap.foreach { case (name, arg) =>
+      val position = nameToPositionMap(name)
+      val param = params(position)
+      val paramType = param.dataType
+      val argType = arg.expr.dataType
+      if (paramType != argType) {
+        throw new AnalysisException(s"Wrong arg type for '${param.name}': expected $paramType but got $argType")
+      }
+      values(position) = toScalaValue(arg.expr)
+    }
+
+    // assign default values for optional params
+    params.foreach {
+      case p: OptionalParameter if !nameToArgMap.contains(p.name) =>
+        val position = nameToPositionMap(p.name)
+        values(position) = p.defaultValue
+      case _ =>
+    }
+
+    values
+  }
+
+  private def buildNameToArgMapUsingNames(
+      args: Seq[CallArgument],
+      nameToPositionMap: Map[String, Int]): Map[String, CallArgument] = {
+
+    val nameToArgMap = mutable.LinkedHashMap.empty[String, CallArgument]
+    args.foreach { arg =>
+      val name = arg.name.get
+
+      if (nameToArgMap.contains(name)) {
+        throw new AnalysisException(s"Duplicate procedure argument: '$name'")
+      }
+
+      if (!nameToPositionMap.contains(name)) {
+        throw new AnalysisException(s"Unknown argument name: '$name'")
+      }
+
+      nameToArgMap.put(name, arg)
+    }
+    nameToArgMap.toMap
+  }
+
+  private def buildNameToArgMapUsingPositions(
+      args: Seq[CallArgument],
+      params: Seq[Parameter]): Map[String, CallArgument] = {
+
+    val nameToArgMap = mutable.LinkedHashMap.empty[String, CallArgument]
+    args.zipWithIndex.foreach { case (arg, position) =>
```

Comment:
I think this is another one where you can do 
Map toMap, and maybe do the error checking up front. Ie args.length > params.size, rather than in the loop

Reply:
Will update.

Reply:
Could be easily updated if the order is not a concern.

```
    if (args.size > params.size) {
      throw new AnalysisException("Too many arguments for procedure")
    }

    args.zipWithIndex.map { case (arg, position) =>
      val param = params(position)
      param.name -> arg
    }.toMap
```

Reply:
If order is a concern we should just keep it as a list of tuples